## 1.1.差异表达分析

### 1) 数据说明

shape-map测序结果中不加药的样本可以作为RNAse样本。这里我们以liulab batch4数据为例进行说明。
本节分为两个步骤：Mapping和Differential expression。

Mapping参考文件为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping。整体流程脚本可参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/script/01.fasrQC-star-lxf.sh。

Differential expression参考文件为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf。整体流程脚本可参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/05-1.Deseq2.R。


### 2) Mapping

注意这里我们采用了**Arabidopsis_thaliana.TAIR10.34.gtf**的注释文件。

```sh
rRNA_bowtie_index=$"/BioII/lulab_b/liuxiaofan/database/ATH/index/bowtie1/rRNA/Arabidopsis_thaliana.TAIR10.34.rRNA"
GTF=$"/BioII/lulab_b/liuxiaofan/database/ATH/GTF/Arabidopsis_thaliana.TAIR10.34.gtf"
STAR_genome_index=$"/BioII/lulab_b/liuxiaofan/database/ATH/index/STAR/genome"
```

#### 2a\) 导入部分

```sh
#!/bin/bash
#SBATCH -J control.CD1_1
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --output=/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/log/out/control.CD1_1.stdout_lxf.log
#SBATCH --error=/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/log//err/control.CD1_1.stderr_lxf.log
```
#### 2b\) fastqc

```sh
echo start fastqc
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/
mkdir 0.fastqc
cd 0.fastqc
mkdir control
cd control
mkdir CD1_1

/BioII/lulab_b/chenyinghui/software/FastQC/fastqc /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/data//control/CD1_1_S4_L003_R1_001.fastq.gz /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/data//control/CD1_1_S4_L003_R2_001.fastq.gz --outdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//0.fastqc/control/CD1_1 --noextract >> /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/data//CD1_1.log 2>&1

echo fastqc success
```

#### 2c\) trimmed

```sh
echo start trimmed
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/
mkdir 1.quality_control
cd 1.quality_control
mkdir control
cd control
mkdir CD1_1
cd CD1_1

/BioII/lulab_b/chenyinghui/software/fastp/fastp -i /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/data//control/CD1_1_S4_L003_R1_001.fastq.gz -I /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/data//control/CD1_1_S4_L003_R2_001.fastq.gz -o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//1.quality_control/control/CD1_1/CD1_1.clean.1.fastq.gz -O /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//1.quality_control/control/CD1_1/CD1_1.clean.2.fastq.gz --thread=4 -l 15 -j CD1_1.json -h CD1_1.html

echo trimmed success
```

#### 2d\) remove_rRNA

```sh
echo start remove_rRNA
mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA
mkdir fastq
cd fastq
mkdir control
cd control
mkdir CD1_1

cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA
mkdir rRNA
cd rRNA
mkdir control
cd control
mkdir CD1_1

#source /BioII/lulab_b/containers/singularity/wrappers/bashrc
/BioII/lulab_b/chenyinghui/software/bowtie/bowtie-1.2.3/bowtie -n 0 -y -a --norc --best --strata -S -p 4 -l 15 --un /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.rm_rRNA.fq /BioII/lulab_b/liuxiaofan/database/ATH/index/bowtie1/rRNA/Arabidopsis_thaliana.TAIR10.34.rRNA -1 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//1.quality_control/control/CD1_1/CD1_1.clean.1.fastq.gz -2 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//1.quality_control/control/CD1_1/CD1_1.clean.2.fastq.gz /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.alngned_rRNA.txt

gzip /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.rm_rRNA_1.fq
gzip /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.rm_rRNA_2.fq
rm /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.alngned_rRNA.txt

echo remove_rRNA success
```

#### 2e\) mapping

```sh
echo start mapping

mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf
mkdir control
cd control
mkdir CD1_1
cd CD1_1

/BioII/lulab_b/chenyinghui/software/STAR/STAR-2.7.3a/bin/Linux_x86_64_static/STAR \
--runThreadN 8 \
--limitBAMsortRAM 20000000000 \
--outFilterType BySJout \
--outFilterMismatchNmax 10  \
--genomeDir /BioII/lulab_b/liuxiaofan/database/ATH/index/STAR/genome \
--readFilesIn /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.rm_rRNA_1.fq.gz /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.rm_rRNA_2.fq.gz \
--readFilesCommand 'zcat' \
--outFileNamePrefix CD1_1  \
--outSAMtype BAM Unsorted \
--quantMode TranscriptomeSAM GeneCounts \
--outSAMattributes All  --outSAMstrandField intronMotif --outBAMcompression 6 --outReadsUnmapped Fastx

/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools sort -T /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.out.sorted -o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam  /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.out.bam
/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools sort -T /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.toTranscriptome.out.sorted -o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.toTranscriptome.out.sorted.bam  /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.toTranscriptome.out.bam

/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools index /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam
/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools index  /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.toTranscriptome.out.sorted.bam

echo mapping success
```

这里注意
* outFilterMismatchNmax=10，是因为考虑突变后错配较多
* limitBAMsortRAM=20000000000，outSAMtype BAM Unsorted，是因为sort时内存溢出错误，利用samtools单独进行排序

#### 2f\) Read counts
```sh
echo start read_counts

mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/
mkdir control
cd control
mkdir CD1_1
cd CD1_1

/Share2/home/lulab/maziyuan/app/subread-2.0.0-Linux-x86_64/bin/featureCounts \
-T 8 \
-s 2 \
-p -t CDS \
-g gene_id \
-a /BioII/lulab_b/liuxiaofan/database/ATH/GTF/Arabidopsis_thaliana.TAIR10.34.gtf \
-o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/control/CD1_1/CD1_1.featurecounts.txt /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam

/Share2/home/lulab/maziyuan/app/subread-2.0.0-Linux-x86_64/bin/featureCounts \
-T 8 \
-s 2 \
-p -t exon \
-g gene_id \
-a /BioII/lulab_b/liuxiaofan/database/ATH/GTF/Arabidopsis_thaliana.TAIR10.34.gtf \
-o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/control/CD1_1/CD1_1.featurecounts.all.txt /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam

echo read_counts success

#merge
mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/result
echo -e 'gene_id 	 CD1_1' > /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/result/CD1_1.txt
cat /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/control/CD1_1/CD1_1.featurecounts.txt | grep -v '#' | grep -v 'Geneid' | cut -f 1,7 >> /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/result/CD1_1.txt

echo -e 'gene_id 	 CD1_1' > /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/result/CD1_1.all.txt
cat /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/control/CD1_1/CD1_1.featurecounts.all.txt | grep -v '#' | grep -v 'Geneid' | cut -f 1,7 >> /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//4.read_count_lxf/result/CD1_1.all.txt
echo success
```

这里featureCounts的-t参数指定reads统计区域，我们分别统计exon和CDS两个类型，得到CD1_1.all.txt，CD1_1.txt文件。统计exon的方法（CD1_1.all.txt）用于差异表达分析，统计CDS的方法（CD1_1.txt）用于差异翻译分析（与Riboseq数据统计方法统一）。

### 3\) Differential expression

我们分别对WT_UV-，WT_UV+，UVR8_UV-，UVR8_UV+条件下的3个样本进行分析，分析加光前后的差异表达基因。

#### 3a\) Count matrix

python脚本

```python
import numpy as np
import pandas as pd
import re
import os

WT = ['C1_1','C1_2','C1_3','C0_1','C0_2','C0_3']
sample_WT=['CD1_1','CD1_2','CD1_3','CD0_1','CD0_2','CD0_3']
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/'
data_list=[]
for i in range(len(WT)):
	print (WT[i])
	temp = pd.read_csv(path + WT[i]+'/mapping/4.read_count_lxf/result/'+sample_WT[i]+'.all.txt', sep='\t',index_col=0)
	# print(temp)
	#temp.set_index("gene_id",inplace=True)
	data_list.append(temp)
matrix_WT_C = data_list[0]
for i in range(len(data_list)-1):
	matrix_WT_C = matrix_WT_C.join(data_list[i+1],how="outer")
matrix_WT_C = matrix_WT_C.sort_index(axis = 1)

UVR = ['U1_1','U1_2','U1_3','U0_1','U0_2','U0_3']
sample_UVR=['UD1_1','UD1_2','UD1_3','UD0_1','UD0_2','UD0_3']
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/'
data_list=[]
for i in range(len(UVR)):
	print(UVR[i])
	temp = pd.read_csv(path + UVR[i]+'/mapping/4.read_count_lxf/result/'+sample_UVR[i]+'.all.txt', sep='\t',index_col=0)
	# print(temp)
	#temp.set_index("gene_id",inplace=True)
	data_list.append(temp)
matrix_UVR_C = data_list[0]
for i in range(len(data_list)-1):
	matrix_UVR_C = matrix_UVR_C.join(data_list[i+1],how="outer")
matrix_UVR_C = matrix_UVR_C.sort_index(axis = 1)


matrix = pd.concat([matrix_WT_C,matrix_UVR_C],axis=1)
matrix.to_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_all.txt',sep='\t',index=True,header=True)

```

#### 3b\) DESeq2

R脚本

```R
#! /usr/lib/R/bin/Rscript --vanilla
library(DESeq2)
raw_count <- read.table("/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_all.txt",sep='\t',header=T)

#col0
col0_raw_count <- raw_count[c("gene_id","CD1_1","CD1_2","CD1_3","CD0_1","CD0_2","CD0_3")]
row.names(col0_raw_count) <- col0_raw_count[,1]
col0_raw_count <- col0_raw_count[,-1]
col0_raw_count <- col0_raw_count[rowSums(col0_raw_count)>100,]

countData_col0 <- col0_raw_count
condition_merge <- factor(c("control","control","control","KD","KD","KD"))
#batch <- factor(c("batch1","batch1","batch1","batch1","batch1","batch1"))
colData <- data.frame(row.names=colnames(countData_col0),condition_merge)

dds <- DESeqDataSetFromMatrix(countData_col0,colData,design=~condition_merge)
dds2 <- DESeq(dds)
res <- results(dds2)

res <- res[order(res$padj),]
diff_gene_deseq_col0 <- subset(res,padj<0.05&(log2FoldChange>1|log2FoldChange< -1))
col0_gene_names <- row.names(diff_gene_deseq_col0)
write.csv(col0_gene_names,"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/DESeq2/wt/wt_gene_list.txt",sep='\t',row.names=F,quote=F)
resdata_col0 <- merge(as.data.frame(res),as.data.frame(counts(dds)),by="row.names",sort=FALSE)
write.csv(resdata_col0,"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/DESeq2/wt/wt_rawdata.csv",sep='\t',row.names=F,quote=F)


#coil
raw_count <- read.table("/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_all.txt",sep='\t',header=T)
coil_raw_count <- raw_count[c("gene_id","UD1_1","UD1_2","UD1_3","UD0_1","UD0_2","UD0_3")]
row.names(coil_raw_count) <- coil_raw_count[,1]
coil_raw_count <- coil_raw_count[,-1]
coil_raw_count <- coil_raw_count[rowSums(coil_raw_count)>100,]

countData_coil <- coil_raw_count
condition_merge <- factor(c("control","control","control","KD","KD","KD"))
#batch <- factor(c("batch1","batch1","batch1","batch1","batch1","batch1"))
colData <- data.frame(row.names=colnames(countData_coil),condition_merge)

dds <- DESeqDataSetFromMatrix(countData_coil,colData,design=~condition_merge)
dds2 <- DESeq(dds)
res <- results(dds2)

res <- res[order(res$padj),]
diff_gene_deseq_coil <- subset(res,padj<0.05&(log2FoldChange>1|log2FoldChange< -1))
coil_gene_names <- row.names(diff_gene_deseq_coil)
write.csv(coil_gene_names,"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/DESeq2/uvr8/uvr8_gene_list.txt",sep='\t',row.names=F,quote=F)
resdata_coil <- merge(as.data.frame(res),as.data.frame(counts(dds)),by="row.names",sort=FALSE)
write.csv(resdata_coil,"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/DESeq2/uvr8/uvr8_rawdata.csv",sep='\t',row.names=F,quote=F)
```

#### 3c\) edgeR

R脚本

```R
#! /usr/lib/R/bin/Rscript --vanilla
library(edgeR)

raw_count <- read.table("/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_all.txt",sep='\t',header=T)


#raw <- read.table(input,header = T, sep="\t", row.names=1)

#col0
col0_raw_count <- raw_count[c("gene_id","CD1_1","CD1_2","CD1_3","CD0_1","CD0_2","CD0_3")]
row.names(col0_raw_count) <- col0_raw_count[,1]
col0_raw_count <- col0_raw_count[,-1]
col0_raw_count <- col0_raw_count[rowSums(col0_raw_count)>100,]

countData_col0 <- col0_raw_count

dgListGroups <- c(rep("Control",3),rep("Treat",3))
dgList <- DGEList(counts=countData_col0, genes=rownames(countData_col0),group=factor(dgListGroups))

dgList <- calcNormFactors(dgList, method="TMM")
countsPerMillion <- cpm(dgList, normalized.lib.sizes=TRUE)
#write.table(countsPerMillion,file=cpmFile,quote=F, sep="\t", row.names=T, col.names=T)

#countCheck <- countsPerMillion > 1
#keep <- which(rowSums(countCheck) >= 5)
#dgList<-dgList[keep,]
design.mat <- model.matrix(~0 + dgList$sample$group)
colnames(design.mat) <- levels(dgList$sample$group)

d2 <- estimateGLMCommonDisp(dgList, design=design.mat)
d2 <- estimateGLMTrendedDisp(d2, design=design.mat)
d2 <- estimateGLMTagwiseDisp(d2, design=design.mat)

fit <- glmFit(d2, design.mat)
lrt <- glmLRT(fit,contrast=c(-1,1))

edgeR_result <- topTags(lrt,n=nrow(dgList))
outfile <- '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/edgeR/wt/wt_rawdata.csv'
write.table(edgeR_result$table,file = outfile,sep = "\t",quote = F,row.names=F,col.names=T)


#coil
raw_count <- read.table("/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_all.txt",sep='\t',header=T)
coil_raw_count <- raw_count[c("gene_id","UD1_1","UD1_2","UD1_3","UD0_1","UD0_2","UD0_3")]
row.names(coil_raw_count) <- coil_raw_count[,1]
coil_raw_count <- coil_raw_count[,-1]
coil_raw_count <- coil_raw_count[rowSums(coil_raw_count)>100,]

countData_coil <- coil_raw_count

dgListGroups <- c(rep("Control",3),rep("Treat",3))
dgList <- DGEList(counts=countData_coil, genes=rownames(countData_coil),group=factor(dgListGroups))

dgList <- calcNormFactors(dgList, method="TMM")
countsPerMillion <- cpm(dgList, normalized.lib.sizes=TRUE)
#write.table(countsPerMillion,file=cpmFile,quote=F, sep="\t", row.names=T, col.names=T)

#countCheck <- countsPerMillion > 1
#keep <- which(rowSums(countCheck) >= 5)
#dgList<-dgList[keep,]
design.mat <- model.matrix(~0 + dgList$sample$group)
colnames(design.mat) <- levels(dgList$sample$group)

d2 <- estimateGLMCommonDisp(dgList, design=design.mat)
d2 <- estimateGLMTrendedDisp(d2, design=design.mat)
d2 <- estimateGLMTagwiseDisp(d2, design=design.mat)

fit <- glmFit(d2, design.mat)
lrt <- glmLRT(fit,contrast=c(-1,1))

edgeR_result <- topTags(lrt,n=nrow(dgList))
outfile <- '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/edgeR/uvr8/uvr8_rawdata.csv'
write.table(edgeR_result$table,file = outfile,sep = "\t",quote = F,row.names=F,col.names=T)
```

### 4\) 结果

最后我们得到的结果主要有以下文件：

```text
exp_lxf
├── DESeq2
      ├── uvr8
           └── uvr8_rawdata.csv
     └── wt
          └── wt_rawdata.csv
├── edgeR
      ├── uvr8
           └── uvr8_rawdata.csv
     └── wt
          └── wt_rawdata.csv
```

这里的文件都是没有根据FDR过滤的，我们认为其中FDR<0.05的基因为差异表达基因。
