## 1.2.差异剪切分析

### 1\) 数据说明

shape-map测序结果中不加药的样本可以作为RNAse样本。这里我们以liulab batch4数据为例进行说明。
本节分为两个步骤：Mapping和Differential splicing。其中Mapping步骤与1.1.差异表达分析基本一致，其中STAR参数有所不同。

Mapping参考文件为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping。整体流程脚本可参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/script/01.fasrQC-star-lxf.sh。

Differential splicing参考文件为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf。整体流程脚本可参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/wt.UVB-vs-noUVB.sh和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/uvr8.UVB-vs-noUVB.sh。

### 2\) Mapping

注意这里fastqc，trimmed，remove_rRNA与1.1差异表达中一致，我们利用前面得到的clean fasta文件进行mapping。
注意这里我们采用了**Arabidopsis_thaliana.TAIR10.34.gtf**的注释文件。

```sh
rRNA_bowtie_index=$"/BioII/lulab_b/liuxiaofan/database/ATH/index/bowtie1/rRNA/Arabidopsis_thaliana.TAIR10.34.rRNA"
GTF=$"/BioII/lulab_b/liuxiaofan/database/ATH/GTF/Arabidopsis_thaliana.TAIR10.34.gtf"
STAR_genome_index=$"/BioII/lulab_b/liuxiaofan/database/ATH/index/STAR/genome"
```


```sh
echo start mapping
mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf_splicing
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf_splicing
mkdir control
cd control
mkdir CD1_1
cd CD1_1
/BioII/lulab_b/chenyinghui/software/STAR/STAR-2.7.3a/bin/Linux_x86_64_static/STAR \
--runThreadN 8 \
--limitBAMsortRAM 20000000000 \
--outFilterMismatchNmax 10  \
--genomeDir /BioII/lulab_b/liuxiaofan/database/ATH/index/STAR/genome \
--readFilesIn /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.rm_rRNA_1.fq.gz /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//2.remove_rRNA/fastq/control/CD1_1/CD1_1.rm_rRNA_2.fq.gz \
--readFilesCommand 'zcat' \
--outFileNamePrefix CD1_1  \
--outSAMtype BAM Unsorted \
--quantMode TranscriptomeSAM GeneCounts \
--outSAMattributes All  --outSAMstrandField intronMotif --outBAMcompression 6 --outReadsUnmapped Fastx

/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools sort -T /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf_splicing/control/CD1_1/CD1_1Aligned.out.sorted -o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf_splicing/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam  /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf_splicing/control/CD1_1/CD1_1Aligned.out.bam


/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools index /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping//3.mapping_lxf_splicing/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam

echo mapping success
```

这里注意
* 与差异表达不同的地方主要是掉outFilterType BySJout参数
* outFilterMismatchNmax=10，是因为考虑突变后错配较多
* limitBAMsortRAM=20000000000，outSAMtype BAM Unsorted，是因为sort时内存溢出错误，利用samtools单独进行排序

### 3\) Differential splicing

可参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf下脚本进行。

#### 3a\) 利用rMATS计算splicing events

脚本：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/wt.UVB-vs-noUVB.sh和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/uvr8.UVB-vs-noUVB.sh

```sh
#!/bin/bash
#SBATCH -J LUAD
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --output=LUAD.out
#SBATCH --error=LUAD.err

echo rMATS start `date`
export PATH=/BioII/lulab_b/chenyinghui/software/samtools/samtools-1.9:$PATH
export PATH=/BioII/lulab_b/chenyinghui/software/conda2/bin:/apps/anaconda2/bin:$PATH
export LD_LIBRARY_PATH=/BioII/lulab_b/chenyinghui/software/conda2/lib:/BioII/lulab_b/chenyinghui/software/conda2/lib64:/BioII/lulab_b/chenyinghui/software/conda3/lib:/BioII/lulab_b/chenyinghui/software/conda3/lib64:/BioII/lulab_b/chenyinghui/software/glibc-build:/BioII/lulab_b/chenyinghui/software/liblzma/lib:$LD_LIBRARY_PATH

/BioII/lulab_b/chenyinghui/software/conda2/bin/python2 /BioII/lulab_b/chenyinghui/software/rMATS/rmats_turbo_v4_1_0_python_2_7/rmats.py --b1 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/wt.noUVB.txt --b2 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/wt.UVB.txt --gtf /BioII/lulab_b/liuxiaofan/database/ATH/GTF/Arabidopsis_thaliana.TAIR10.34.gtf --od /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/wt.UVB-vs-noUVB -t paired --readLength 151 --cstat 0.0001 --tmp /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/wt.UVB-vs-noUVB/tmp --nthread 4 --variable-read-length

echo rMATS end `date`
```

其中wt.noUVB.txt和wt.UVB.txt文件为mapping后bam文件路径。值得注意的是**readLength=151**需要根据输入数据进行给定参数，而**不是固定参数**。

wt.noUVB.txt

```sh
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/3.mapping_lxf_splicing/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_2/mapping/3.mapping_lxf_splicing/control/CD1_2/CD1_2Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_3/mapping/3.mapping_lxf_splicing/control/CD1_3/CD1_3Aligned.sortedByCoord.out.bam
```
wt.UVB.txt

```sh
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C0_1/mapping/3.mapping_lxf_splicing/control/CD0_1/CD0_1Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C0_2/mapping/3.mapping_lxf_splicing/control/CD0_2/CD0_2Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C0_3/mapping/3.mapping_lxf_splicing/control/CD0_3/CD0_3Aligned.sortedByCoord.out.bam
```

#### 3b\) splicing events结果处理

* rMATS结果过滤(psi > 0.1 & pvalue < 0.05)

脚本：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/1.splice_sig_psi.sh

```sh
mkdir wt.UVB-vs-noUVB_filtered_psi_0.1
mkdir uvr8.UVB-vs-noUVB_filtered_psi_0.1
#p value<0.05 & psi>0.1
python bin/splice_sig_psi.py PValue 0.05 0.1 wt.UVB-vs-noUVB wt.UVB-vs-noUVB_filtered_psi_0.1
python bin/splice_sig_psi.py PValue 0.05 0.1 uvr8.UVB-vs-noUVB uvr8.UVB-vs-noUVB_filtered_psi_0.1
rename txt xls wt.UVB-vs-noUVB_filtered_psi_0.1/*txt
rename txt xls uvr8.UVB-vs-noUVB_filtered_psi_0.1/*txt
```
 
 输出目录： uvr8.UVB-vs-noUVB_filtered_psi_0.1 , wt.UVB-vs-noUVB_filtered_psi_0.1

*  对过滤后differential splicing gene进行通路注释

脚本：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/2.clusterProfiler.sh和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/3.GO_anno_all.sh

输出结果：uvr8.UVB-vs-noUVB_filtered_psi_0.1中*.MATS.JCEC.annot.csv , wt.UVB-vs-noUVB_filtered_psi_0.1中*.MATS.JCEC.annot.csv

* 提取出splicing区域的起始、终止位置信息

脚本：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/4.grep_region.sh

 输出目录： uvr8.UVB-vs-noUVB_filtered_psi_0.1/region , wt.UVB-vs-noUVB_filtered_psi_0.1/region
 
 #### 3c\) splicing events批量画图
 
 脚本：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/rmats2sashimiplot.wt.UVB-vs-noUVB.all.sh和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/rmats2sashimiplot-uvr8.UVB-vs-noUVB.all.sh
 
 ```sh
inputDir="/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/splicing_lxf/wt.UVB-vs-noUVB_filtered_psi_0.1/wt.UVB-vs-noUVB_filtered_psi_0.1"
outDir="$inputDir/wt.UVB-vs-noUVB_sashimiplot_all"
for event in "SE" "A5SS" "A3SS" "MXE" "RI"
do
mkdir -p $outDir/$event
source /BioII/lulab_b/containers/singularity/wrappers/bashrc
export PATH=/BioII/lulab_b/chenyinghui/software/samtools/samtools-1.9:$PATH
export PATH=/BioII/lulab_b/chenyinghui/software/conda2/bin:/apps/anaconda2/bin:$PATH
export LD_LIBRARY_PATH=/BioII/lulab_b/chenyinghui/software/conda2/lib:/BioII/lulab_b/chenyinghui/software/conda2/lib64:/BioII/lulab_b/chenyinghui/software/conda3/lib:/BioII/lulab_b/chenyinghui/software/conda3/lib64:/BioII/lulab_b/chenyinghui/software/glibc-build:/BioII/lulab_b/chenyinghui/software/liblzma/lib:$LD_LIBRARY_PATH

/BioII/lulab_b/chenyinghui/software/conda2/bin/rmats2sashimiplot \
--b1 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_1/mapping/3.mapping_lxf_splicing/control/CD1_1/CD1_1Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_2/mapping/3.mapping_lxf_splicing/control/CD1_2/CD1_2Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C1_3/mapping/3.mapping_lxf_splicing/control/CD1_3/CD1_3Aligned.sortedByCoord.out.bam \
--b2 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C0_1/mapping/3.mapping_lxf_splicing/control/CD0_1/CD0_1Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C0_2/mapping/3.mapping_lxf_splicing/control/CD0_2/CD0_2Aligned.sortedByCoord.out.bam,/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/C0_3/mapping/3.mapping_lxf_splicing/control/CD0_3/CD0_3Aligned.sortedByCoord.out.bam \
--l1 Control_wt_noUVB \
--l2 Treat_wt_UVB \
--exon_s 1 \
--intron_s 2 \
--min-counts 0 \
-t $event \
-e $inputDir/${event}.MATS.JCEC.all.sashimiplot.txt \
-o $outDir/$event

done
rm -r $outDir/*/Sashimi_index*
```

其中{event}.MATS.JCEC.all.sashimiplot.txt是需要绘图的不同事件的信息，如果需要部分画图可对该文件进行筛选。

输出目录为wt.UVB-vs-noUVB_filtered_psi_0.1/wt.UVB-vs-noUVB_sashimiplot_all和uvr8.UVB-vs-noUVB_filtered_psi_0.1/uvr8.UVB-vs-noUVB_sashimiplot_all。其中每个文件为画图结果，开头ID与{event}.MATS.JCEC.all.sashimiplot.txt文件第一列ID相对应。


