## 2.Riboseq数据分析

本节主要介绍Riboseq数据的分析流程和方法，分为Mapping，周期性和ORF分析，差异翻译分析三个部分。

### 1) 数据说明

我们以liulab的第二批Riboseq数据为例进行说明。具体内容可参考文件夹/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq。

Mapping的整体流程脚本可参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/script/01.pipline_lxf.sh。


### 2) Mapping

注意这里我们采用了**Arabidopsis_thaliana.TAIR10.34.gtf**的注释文件。

```sh
rRNA_bowtie_index=$"/BioII/lulab_b/liuxiaofan/database/ATH/index/bowtie1/rRNA/Arabidopsis_thaliana.TAIR10.34.rRNA"
GTF=$"/BioII/lulab_b/liuxiaofan/database/ATH/GTF/Arabidopsis_thaliana.TAIR10.34.gtf"
STAR_genome_index=$"/BioII/lulab_b/liuxiaofan/database/ATH/index/STAR/genome"
```

#### 2a\) 导入部分

```sh
#!/bin/bash
#BSUB -J wtnouvb1
#BSUB -o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//log/out/wtnouvb1.stdout_3.log
#BSUB -e /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//log/err/wtnouvb1.stderr_3.log
#BSUB -R span[hosts=1]
#BSUB -n 4
#BSUB -q Z-LU
```
#### 2b\) fastqc

```sh
echo start fastqc
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/
mkdir 0.fastqc
cd 0.fastqc
mkdir wtnouvb1
cd wtnouvb1
/BioII/lulab_b/chenyinghui/software/FastQC/fastqc -t 4 /Share2/home/lulab1/Riboshape/riboshape_liulab_ribo_batch4/raw//wtnouvb1.fq.gz --outdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//0.fastqc/wtnouvb1 --noextract
echo fastqc success
```

#### 2c\) trimmed

```sh
echo start trimmed

cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/
mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//1.quality_control
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//1.quality_control

mkdir wtnouvb1
cd wtnouvb1
/BioII/lulab_b/chenyinghui/software/fastp/fastp --length_limit 50 --adapter_fasta /BioII/lulab_b/chenyinghui/software/fastp/adapters.fa -i /Share2/home/lulab1/Riboshape/riboshape_liulab_ribo_batch4/raw//wtnouvb1.fq.gz -o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//1.quality_control/wtnouvb1/wtnouvb1.clean.fq.gz --thread=4 -l 15 -j wtnouvb1.json -h wtnouvb1.html
mv /BioII/lulab_b/chenyinghui/project/shapeMap/liulab_ribo_seq_2/02.fastp/script/uvr8no1.html /BioII/lulab_b/chenyinghui/project/shapeMap/liulab_ribo_seq_2/02.fastp/
mv /BioII/lulab_b/chenyinghui/project/shapeMap/liulab_ribo_seq_2/02.fastp/script/uvr8no1.json /BioII/lulab_b/chenyinghui/project/shapeMap/liulab_ribo_seq_2/02.fastp/

echo trimmed success
```

#### 2d\) remove_rRNA

```sh
echo start remove_rRNA

mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA
mkdir fastq
cd fastq
mkdir wtnouvb1
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA
mkdir rRNA
cd rRNA
mkdir wtnouvb1

source /BioII/lulab_b/containers/singularity/wrappers/bashrc
/BioII/lulab_b/chenyinghui/software/bowtie/bowtie-1.2.3/bowtie -y -a --norc --best --strata -S -p 4 -l 15 --un /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA/fastq/wtnouvb1/wtnouvb1.rm_rRNA.fq /BioII/lulab_b/liuxiaofan/database/ATH/index/bowtie1/rRNA/Arabidopsis_thaliana.TAIR10.34.rRNA -q /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_riboseq_batch2/02.fastp/wtnouvb1.clean.fq.gz  /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA/fastq/wtnouvb1/wtnouvb1.alngned_rRNA.txt
gzip /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA/fastq/wtnouvb1/wtnouvb1.rm_rRNA.fq
rm /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA/fastq/wtnouvb1/wtnouvb1.alngned_rRNA.txt

echo remove_rRNA success
```

#### 2e\) mapping

```sh
echo start mapping

mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping
mkdir wtnouvb1
cd wtnouvb1

/BioII/lulab_b/chenyinghui/software/STAR/STAR-2.7.3a/bin/Linux_x86_64_static/STAR \
--runThreadN 4 \
--outFilterType BySJout \
--outFilterMismatchNmax 2 \
--outFilterMultimapNmax 1 \
--genomeDir /BioII/lulab_b/liuxiaofan/database/ATH/index/STAR/genome \
--readFilesIn /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//2.remove_rRNA/fastq/wtnouvb1/wtnouvb1.rm_rRNA.fq.gz \
--readFilesCommand 'zcat' \
--outFileNamePrefix  wtnouvb1. \
--outSAMtype BAM SortedByCoordinate \
--quantMode TranscriptomeSAM GeneCounts \
--outSAMattributes All \
--outSAMattrRGline ID:1 LB:ribo_seq PL:ILLUMINA SM:wtnouvb1 \
--outBAMcompression 6 \
--outReadsUnmapped Fastx

/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools sort -T /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping/wtnouvb1/wtnouvb1.Aligned.toTranscriptome.out.sorted -o /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping/wtnouvb1/wtnouvb1.Aligned.toTranscriptome.out.sorted.bam /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping/wtnouvb1/wtnouvb1.Aligned.toTranscriptome.out.bam

/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools index  /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping/wtnouvb1/wtnouvb1.Aligned.toTranscriptome.out.sorted.bam
/BioII/lulab_b/chenyinghui/software/conda3/bin/samtools index /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping/wtnouvb1/wtnouvb1.Aligned.sortedByCoord.out.bam

echo mapping success
```

这里注意
* outFilterMultimapNmax=1，是因为不考虑Multimap reads


#### 2f\) Read counts
```sh
echo start read

mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//4.read_count_HTSeq
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//4.read_count_HTSeq
mkdir wtnouvb1
cd wtnouvb1

/BioII/lulab_b/chenyinghui/software/conda3/bin/htseq-count -f bam -s no -i gene_id -t CDS -m union /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//3.mapping/wtnouvb1/wtnouvb1.Aligned.sortedByCoord.out.bam /BioII/lulab_b/liuxiaofan/database/ATH/GTF/Arabidopsis_thaliana.TAIR10.34.gtf > /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//4.read_count_HTSeq/wtnouvb1.read_count.HTSeq.txt

grep __ /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//4.read_count_HTSeq/wtnouvb1.read_count.HTSeq.txt > /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//4.read_count_HTSeq/wtnouvb1.read_count.HTSeq.txt.summary
sed -i '/^__/d' /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq//4.read_count_HTSeq/wtnouvb1.read_count.HTSeq.txt
echo read success
```

### 3\) 周期性和ORF分析

* 获得周期性分析报告

脚本：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/script/06.RiboCode_metaplot_lxf.sh

```sh
dataPath='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/3.mapping'
outDir='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/06.RiboCode/metaplot'
sample_list='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_riboseq_batch2/sample_list.txt'
RiboCode_annot='/BioII/lulab_b/liuxiaofan/database/ATH/RiboCode_annot'
#STAR_genome_index='/BioII/lulab_b/chenyinghui/database/Homo_sapiens/GRCh38/GRCh38_STAR_index'

script="$outDir/script"
if [ ! -d $script ]
then
mkdir -p $script
fi

for i in `cat $sample_list`
do

echo -e "#!/bin/bash
#BSUB -J \"metaplot\"
#BSUB -o ${i}.metaplot.stdout.log
#BSUB -e ${i}.metaplot.stderr.log
#BSUB -R \"span[hosts=1]\"
#BSUB -n 1
#BSUB -q Z-LU
echo start \`date\`

if [ ! -d $outDir/${i} ]
then
mkdir -p $outDir/${i}
fi
export PATH=/BioII/lulab_b/liuxiaofan/software/conda3/bin:\$PATH
metaplots -a $RiboCode_annot -r $dataPath/${i}/${i}.Aligned.toTranscriptome.out.bam -o $outDir/$i. -m 26 -M 34 -s no -pv1 0.05 -pv2 0.05

echo end \`date\`
" > "$script/${i}.metaplots.sh"
done
```
其中-m 26， -M 34为最小值和最大值
输出目录：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/06.RiboCode/metaplot

* 获得ORF区域结果

脚本：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/script/06.RiboCode_ORF.sh

```sh
dataPath='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/06.RiboCode'
outDir='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/06.RiboCode'
sample_list='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_riboseq_batch2/sample_list.txt'
RiboCode_annot="/BioII/lulab_b/liuxiaofan/database/ATH/RiboCode_annot"
script="$outDir/script"
if [ ! -d $script ]
then
mkdir -p $script
fi

for i in `cat $sample_list`
do
echo -e "
#!/bin/bash
#BSUB -J \"RiboCode\"
#BSUB -o $i.stdout.log
#BSUB -e $i.stderr.log
#BSUB -n 1
#BSUB -q Z-LU

echo start \`date\`

mkdir $outDir/$i
cd $outDir/$i

export PATH=/BioII/lulab_b/liuxiaofan/software/conda3/bin:\$PATH

RiboCode -a $RiboCode_annot -c $outDir/metaplot/${i}._pre_config.txt -l no -g -o $outDir/$i

echo end \`date\`
" > "$script/${i}.RiboCode.sh"
done
```

输出目录为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/06.RiboCode。wtno1_collapsed.txt为ORF结果。


### 4\) Differential translation efficiency

我们利用Xtail基于Riboseq的count matrix和RNAseq的count matrix计算差异翻译效率。为了方便对比RNAseq只统计CDS区域的reads。

#### 4a\) RNAseq汇总

python脚本

```python
import numpy as np
import pandas as pd
import re
import os

WT = ['C1_1','C1_2','C1_3','C0_1','C0_2','C0_3']
sample_WT=['CD1_1','CD1_2','CD1_3','CD0_1','CD0_2','CD0_3']
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/'
data_list=[]
for i in range(len(WT)):
	print (WT[i])
	temp = pd.read_csv(path + WT[i]+'/mapping/4.read_count_lxf/result/'+sample_WT[i]+'.txt', sep='\t',index_col=0)
	# print(temp)
	#temp.set_index("gene_id",inplace=True)
	data_list.append(temp)
matrix_WT_C = data_list[0]
for i in range(len(data_list)-1):
	matrix_WT_C = matrix_WT_C.join(data_list[i+1],how="outer")
matrix_WT_C = matrix_WT_C.sort_index(axis = 1)

UVR = ['U1_1','U1_2','U1_3','U0_1','U0_2','U0_3']
sample_UVR=['UD1_1','UD1_2','UD1_3','UD0_1','UD0_2','UD0_3']
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/'
data_list=[]
for i in range(len(UVR)):
	print(UVR[i])
	temp = pd.read_csv(path + UVR[i]+'/mapping/4.read_count_lxf/result/'+sample_UVR[i]+'.txt', sep='\t',index_col=0)
	# print(temp)
	#temp.set_index("gene_id",inplace=True)
	data_list.append(temp)
matrix_UVR_C = data_list[0]
for i in range(len(data_list)-1):
	matrix_UVR_C = matrix_UVR_C.join(data_list[i+1],how="outer")
matrix_UVR_C = matrix_UVR_C.sort_index(axis = 1)


matrix = pd.concat([matrix_WT_C,matrix_UVR_C],axis=1)
matrix.to_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_CDS.txt',sep='\t',index=True,header=True)

```

#### 4b\) Riboseq汇总

python脚本

```python

import numpy as np
import pandas as pd
import re
import os

# WT = ['C1_1','C1_2','C1_3','C0_1','C0_2','C0_3']
sample_WT=['wtnouvb1','wtnouvb2','wtnouvb3','wtuvb1','wtuvb2','wtuvb3']
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/4.read_count_HTSeq/'
data_list=[]
for i in range(len(sample_WT)):
	print (sample_WT[i])
	temp = pd.read_csv(path +sample_WT[i]+'.read_count.HTSeq.txt', sep='\t',index_col=0,header=None)
	temp.columns=[sample_WT[i]]
	# print(temp)
	#temp.set_index("gene_id",inplace=True)
	data_list.append(temp)
matrix = data_list[0]
for i in range(len(data_list)-1):
	matrix = matrix.join(data_list[i+1],how="outer")
matrix = matrix.sort_index(axis = 1)
matrix.to_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/WT_count.txt',sep='\t',index=True,header=True)

# UVR = ['U1_1','U1_2','U1_3','U0_1','U0_2','U0_3']
sample_UVR=['uvr8no1','uvr8no2','uvr8no3','uvr8uvb1','uvr8uvb2','uvr8uvb3']
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/4.read_count_HTSeq/'
data_list=[]
for i in range(len(sample_UVR)):
	print (sample_UVR[i])
	temp = pd.read_csv(path + sample_UVR[i] + '.read_count.HTSeq.txt', sep='\t', index_col=0, header=None)
	temp.columns = [sample_UVR[i]]
	# print(temp)
	#temp.set_index("gene_id",inplace=True)
	data_list.append(temp)
matrix = data_list[0]
for i in range(len(data_list)-1):
	matrix = matrix.join(data_list[i+1],how="outer")
matrix = matrix.sort_index(axis = 1)
matrix.to_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/UVR_count.txt',sep='\t',index=True,header=True)
```

#### 4c\) Xtail分析

R脚本

```R
library(xtail)

#####WT样本分析######
ribo <- read.table('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/WT_count.txt',header=T, quote='',check.names=F, sep='\t',row.names=1)
mrna <- read.table('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_CDS.txt',header=T, quote='',check.names=F, sep='\t',row.names=1)

ribo <- ribo[,c("wtnouvb2","wtnouvb3","wtuvb1","wtuvb2",	"wtuvb3")]
mrna <- mrna[c("CD1_2","CD1_3","CD0_1","CD0_2","CD0_3")]

condition <- c("control","control","treat","treat","treat")
results <- xtail(mrna,ribo,condition,minMeanCount=1,bins=10000)
results_tab <- resultsTable(results,sort.by="pvalue.adjust",log2FCs=TRUE, log2Rs=TRUE)
write.table(results_tab,"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/wt.uvb-vs-nouvb.TE_new.xls",quote=F,sep="\t")

#####UVR8样本分析######
ribo <- read.table('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/UVR_count.txt',header=T, quote='',check.names=F, sep='\t',row.names=1)
mrna <- read.table('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4/exp_lxf/count_CDS.txt',header=T, quote='',check.names=F, sep='\t',row.names=1)

ribo <- ribo[,c("uvr8no1","uvr8no2","uvr8no3","uvr8uvb1","uvr8uvb2",	"uvr8uvb3")]
mrna <- mrna[c("UD1_1","UD1_2","UD1_3","UD0_1","UD0_2","UD0_3")]

condition <- c("control","control","control","treat","treat","treat")
results <- xtail(mrna,ribo,condition,minMeanCount=1,bins=10000)
results_tab <- resultsTable(results,sort.by="pvalue.adjust",log2FCs=TRUE, log2Rs=TRUE)
write.table(results_tab,"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/uvr8.uvb-vs-nouvb.TE.xls",quote=F,sep="\t")
```

输出结果：/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/wt.uvb-vs-nouvb.TE.csv和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq/7.TE_lxf/uvr8.uvb-vs-nouvb.TE.csv，包括所有基因结果，未进行差异过滤。

