## 3.1.Shapemapper重构

本节我们将介绍如何进行Shape-seq分析。首先我们利用Shapemapper软件计算3条转录本，然后利用中间文件(预处理后文件)进行后面的计算，最后得到每个转录本的reactivity文件。

### 1）数据说明

我们以liulab的第四批Shape-seq数据为例进行说明。具体内容可参考文件夹/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_riboseq。

我们选择**计算每个基因最长的转录本**，我们对注释文件、fasta文件进行了预处理，预处理后文件存在/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file。

### 2）Shapemapper软件获取预处理后文件

我们以样本C1_1为例,参考脚本/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/chr1.sh

```sh
#!/bin/bash
#SBATCH -J C1_1_1
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --output=%j.out
#SBATCH --error=%j.err
#source ~/.bashrc

set -e # exit on first error (if any)

# Find the parent folder of this script,
# resolving (possibly nested) symlinks
cd /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/

/data/liuxiaofan/software/shapemap/shapemapper \
--target /BioII/lulab_b/liuxiaofan/database/ATH/tanscript/Arabidopsis_thaliana.TAIR10.34.transcripts_new_1_1.fa \
--name "Ath_C1_1" \
--min-depth 100 \
--min-qual-to-count 20 \
--overwrite \
--modified --folder /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/data/modified/1  \
--untreated --folder /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/data/control/1 \
--star-aligner \
--nproc 8 \
--verbose

cd /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1/
rm -r NormalizeAsGroup
rm -r SplitByTarget_*
rm -r PostAlignment_RNA_*
rm -r AlignPrep/
cd /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1/Modified
rm -r Interleaver/
rm -r Merger/
rm -r ProgressMonitor/
rm -r QualityTrimmer*
rm -r StarAligner/
cd /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1/Untreated
rm -r Interleaver/
rm -r Merger/
rm -r ProgressMonitor/
rm -r QualityTrimmer*
rm -r StarAligner
```
我们得到输出我们得到输出目录/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1，其中Modified/LengthFilter/consolidated.Ath_C1_1_Modified_LengthFilter_filtered.fastq和Untreated/LengthFilter/consolidated.Ath_C1_1_Untreated_LengthFilter_filtered.fastq为后续计算所需的中间文件。

### 3）Mapping

这里我们分别对Modified和Untreated文件进行mapping

* Modified脚本

/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/mapping.sh

```sh
#!/bin/sh
#SBATCH -J C1_1_m
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --output=%j.out
#SBATCH --error=%j.err
source /WORK/Samples/singularity.sh
cd /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1
mkdir shapemapper_temp/Ath_C1_1_new
mkdir shapemapper_temp/Ath_C1_1_new/Modified
mkdir shapemapper_temp/Ath_C1_1_new/Modified/StarAligner
mkdir shapemapper_temp/Ath_C1_1_new/Modified/StarAligner/Ath_C1_1_Modified_StarAligner_logs
rm -r shapemapper_temp/Ath_C1_1_new/Modified/StarAligner/Ath_C1_1_Modified_StarAligner_temp
/BioII/lulab_b/chenyinghui/software/STAR/STAR-2.7.3a/bin/Linux_x86_64_static/STAR --runThreadN 8 --alignEndsType Local --scoreGap -1000000 --scoreDelBase -1 --scoreInsBase -1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 999 --runMode alignReads --outMultimapperOrder Random --outSAMmultNmax 1 --outStd SAM --outSAMattributes MD --genomeDir "/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/STAR_index/StarIndexBuilder_index" --readFilesIn "shapemapper_temp/Ath_C1_1/Modified/LengthFilter/consolidated.Ath_C1_1_Modified_LengthFilter_filtered.fastq" --outTmpDir "shapemapper_temp/Ath_C1_1_new/Modified/StarAligner/Ath_C1_1_Modified_StarAligner_temp" --outFileNamePrefix "shapemapper_temp/Ath_C1_1_new/Modified/StarAligner/Ath_C1_1_Modified_StarAligner_logs"/ >  "shapemapper_temp/Ath_C1_1_new/Modified/StarAligner/Ath_C1_1_Modified_StarAligner_aligned.sam"
```

* Untreated脚本

/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/mapping_un.sh

```sh
#!/bin/sh
#SBATCH -J C1_1_u
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --output=%j.out
#SBATCH --error=%j.err
source /WORK/Samples/singularity.sh
cd /data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1
mkdir shapemapper_temp/Ath_C1_1_new
mkdir shapemapper_temp/Ath_C1_1_new/Untreated
mkdir shapemapper_temp/Ath_C1_1_new/Untreated/StarAligner
mkdir shapemapper_temp/Ath_C1_1_new/Untreated/StarAligner/Ath_C1_1_Untreated_StarAligner_logs
rm -r shapemapper_temp/Ath_C1_1_new/Untreated/StarAligner/Ath_C1_1_Untreated_StarAligner_temp
/BioII/lulab_b/chenyinghui/software/STAR/STAR-2.7.3a/bin/Linux_x86_64_static/STAR --runThreadN 8 --alignEndsType Local --scoreGap -1000000 --scoreDelBase -1 --scoreInsBase -1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 999 --runMode alignReads --outMultimapperOrder Random --outSAMmultNmax 1 --outStd SAM --outSAMattributes MD --genomeDir "/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/STAR_index/StarIndexBuilder_index" --readFilesIn "shapemapper_temp/Ath_C1_1/Untreated/LengthFilter/consolidated.Ath_C1_1_Untreated_LengthFilter_filtered.fastq" --outTmpDir "shapemapper_temp/Ath_C1_1_new/Untreated/StarAligner/Ath_C1_1_Untreated_StarAligner_temp" --outFileNamePrefix "shapemapper_temp/Ath_C1_1_new/Untreated/StarAligner/Ath_C1_1_Untreated_StarAligner_logs"/ >  "shapemapper_temp/Ath_C1_1_new/Untreated/StarAligner/Ath_C1_1_Untreated_StarAligner_aligned.sam"
```

### 4）Split SAM

这里我们需要将mapping得到的SAM文件分割成每个转录本的SAM文件。我们分别对Modified和Untreated文件进行处理。

* Modified脚本

/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/split_sam.sh

```sh
#!/bin/sh
#SBATCH -J C1_1
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --output=%j.out
#SBATCH --error=%j.err
#source ~/.bashrc
sample_list=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/transcript_name_new.txt"
input=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Modified/StarAligner/Ath_C1_1_Modified_StarAligner_aligned.sam"
output_sam=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Modified/SAM/"
rm -r $output_sam
mkdir $output_sam
cd $output_sam
source /WORK/Samples/singularity.sh
echo `date`
/BioII/lulab_b/liuxiaofan/software/conda3/bin/python3 /data/liuxiaofan/project/riboshape/bin2/split_sam.py -i $input -n $sample_list -o $output_sam
echo `date`
echo ok1
cat $input | grep ^[@].* > $output_sam/tail.txt
echo ok
echo `date`
```

* Untreated脚本

/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/split_sam_un.sh

```sh
#!/bin/sh
#SBATCH -J C1_1_un
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --output=%j.out
#SBATCH --error=%j.err
#source ~/.bashrc
sample_list=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/transcript_name_new.txt"
input=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Untreated/StarAligner/Ath_C1_1_Untreated_StarAligner_aligned.sam"

output_sam=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Untreated/SAM/"
rm -r $output_sam
mkdir $output_sam
cd $output_sam
source /WORK/Samples/singularity.sh
echo `date`
/BioII/lulab_b/liuxiaofan/software/conda3/bin/python3 /data/liuxiaofan/project/riboshape/bin2/split_sam.py -i $input -n $sample_list -o $output_sam
echo `date`
echo ok1
cat $input | grep ^[@].* > $output_sam/tail.txt
echo ok
echo `date`
```

这里如果出现内存溢出错误的话，可能是由于mapping后得到的SAM文件太大，如果超过40G容易内存溢出。这里我们可以利用以下脚本对SAM文件进行切割，然后分别进行转录本切割，再将同一个转录本的SAM进行合并。

split -C 30G Ath_C1_1_Untreated_StarAligner_aligned.sam Ath_C1_1_Untreated

### 5）mutation count for each transcript

我们可以对每个转录本进行total reads和mutation reads count统计，为了减少计算时间，我们将所有转录本划分为10份进行同时运算。具体脚本可参考/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/split_merge和/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/split_merge_un文件夹。

我们分别对Modified和Untreated文件进行处理。

* Modified脚本

/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/split_merge/split_merge_1.sh

```sh
#!/bin/sh
#SBATCH -J mC1_1_1
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --output=%j.out
#SBATCH --error=%j.err
#source ~/.bashrc

sample_list=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/transcript_1.txt"
#sample_list=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/test.txt"
output_sam=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Modified/SAM/"
output_MutationParser=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/MutationParser_Modified/"
output_MutationCounter=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/MutationCounter_Modified/"

#mkdir $output_sam

mkdir $output_MutationParser

mkdir $output_MutationCounter

cd $output_sam
source /WORK/Samples/singularity.sh
echo `date`
for transcript in `cat $sample_list`
do
echo $transcript
t=$(echo $transcript| tr -d "\r")
sam_1="$output_sam"/"$t"_1.sam
sam_2="$output_sam"/"$t"_2.sam
sam="$output_sam"/"$t".sam
mutation=${output_MutationParser}/"$t".mutations.mut
count=${output_MutationCounter}/"$t".mutations.txt
len=$(cat tail.txt | grep -P "${t}\t" | grep '@SQ' | awk '{split($3,data,":");print data[2];}')
echo $len

/data/liuxiaofan/software/shapemap/bin/shapemapper_mutation_parser -i ${sam} -o ${mutation} -w -m 10
/data/liuxiaofan/software/shapemap/bin/shapemapper_mutation_counter -i ${mutation} -w -c ${count}  --exclude_3prime 1 --max_internal_match 5 --min_qual 20 --length $len

done
echo `date`
```

* Untreated脚本

/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/split_merge_un/split_merge_un_1.sh

```sh
#!/bin/sh
#SBATCH -J mC1_1_un_1
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --output=%j.out
#SBATCH --error=%j.err
#source ~/.bashrc

sample_list=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/transcript_1.txt"
#sample_list=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/test.txt"
output_sam=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Untreated/SAM/"
output_MutationParser=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/MutationParser_Untreated/"
output_MutationCounter=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/MutationCounter_Untreated/"

#mkdir $output_sam

mkdir $output_MutationParser

mkdir $output_MutationCounter

cd $output_sam
source /WORK/Samples/singularity.sh
echo `date`
for transcript in `cat $sample_list`
do
echo $transcript
t=$(echo $transcript| tr -d "\r")
sam_1="$output_sam"/"$t"_1.sam
sam_2="$output_sam"/"$t"_2.sam
sam="$output_sam"/"$t".sam
mutation=${output_MutationParser}/"$t".mutations.mut
count=${output_MutationCounter}/"$t".mutations.txt
len=$(cat tail.txt | grep -P "${t}\t" | grep '@SQ' | awk '{split($3,data,":");print data[2];}')
echo $len


/data/liuxiaofan/software/shapemap/bin/shapemapper_mutation_parser -i ${sam} -o ${mutation} -w -m 10
/data/liuxiaofan/software/shapemap/bin/shapemapper_mutation_counter -i ${mutation} -w -c ${count}  --exclude_3prime 1 --max_internal_match 5 --min_qual 20 --length $len

done
echo `date`
```

### 5）data summary

我们合并后所有转录本的结果，包括transcript_id,Nucleotide,Modified_mutations,Modified_effective_depth,Untreated_mutations,Untreated_effective_depth, reactivity

脚本为/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/log_test/sum_data_1.sh

```sh
#!/bin/sh
#SBATCH -J C1_1
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --output=%j.out
#SBATCH --error=%j.err
#source ~/.bashrc
Modified_path=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/MutationCounter_Modified/"
Untreated_path=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/MutationCounter_Untreated/"
output_path=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shape_sum/final.modified_unmodified_1.txt"
fa_path=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/fasta.txt"
sample_list=$"/data/liuxiaofan/project/riboshape/3.reference_genome/Ath/target_file/transcript_name_new.txt"
Modified_sam=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Modified/SAM/"
Untreated_sam=$"/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shapemapper_temp/Ath_C1_1_new/Untreated/SAM/"

/BioII/lulab_b/liuxiaofan/software/conda3/bin/python3 /data/liuxiaofan/project/riboshape/bin2/sum_data.py -m ${Modified_path} -u ${Untreated_path} -o ${output_path} -f ${fa_path}

echo `date`

```

### 5）结果

我们最后得到了汇总的结果文件，/data/liuxiaofan/project/riboshape/result_liulab_batch4_2/C1_1/shape_sum/final.modified_unmodified_1.txt。

文件每列内容为transcript_id,Nucleotide,Modified_mutations,Modified_effective_depth,Untreated_mutations,Untreated_effective_depth, reactivity。
